Reshaping Contributing docs (#67785)
WordPress/gutenberg: storybook/stories/contributing/contributing-components.txt

import { Meta } from '@storybook/addon-docs';

<Meta title="Contributing/Components" />

> # Contribute to components

Reviewer: I think it would be better if we reviewed changes as a diff on the [original file](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md), since it's unclear which parts have been changed here from the original.

> Because components are an important part of our design system and it’s WordPress policy to make them backwards compatible, contributions to them are seen as major changes.

Reviewer 2: This sentence feels slightly verbose and uses passive voice. I think something more direct and active like: At WordPress, component contributions are major changes. (Then elaborate if you want about backwards compatibility and maybe link to public info about it?)

Author: Love it — added.

> Because components are an important part of our design system and it’s WordPress policy to make them backwards compatible, contributions to them are seen as major changes.

Reviewer 2: ```suggestion
Because components are an important part of our design system and it’s WordPress policy to make them backward compatible, contributions to them are seen as major changes.
```

> While changing a typo in our documentation can be shipped quite quickly, work on components requires more time and attention from our design community.

Reviewer 2: Instruction opportunity: If you have a small fix like a typo do [this], if you have an enhancement do [that]. (using enhancement as a term because is the GitHub label we use for improvements). I'd also be more explicit about what goes into getting a component contribution in, like for example 'component work requires at least one formal review from a designer'.

Author: Rewrote this section!

*Is this your first contribution? You’ll first need to follow <Link story="Contributing/Get Ready to Contribute">a couple of steps to get started</Link>.*

## Quick Links
- [Creating a New Component](#creating-a-new-component)
- [Extending Existing Components](#extending-existing-components)
- [Code Standards and Best Practices](#code-standards-and-best-practices)
- [Testing and Storybook](#testing-and-storybook)
- [Documentation Guidelines](#documentation-guidelines)

### Creating a New Component

There are 5 steps to add a new component to WPDS.

Step 1: [Do we need this component?](#step-1)
Step 2: [Open a GitHub issue](#step-2) 
Step 3: [Implement your change](#step-3) 
Step 4: [Refinement](#step-4) 
Step 5: [Publish and Maintain](#step-5) 

#### Step 1: Do we need this component? {#step-1}

It’s important to ask yourself this question before starting your contribution. Our library should include components that are generic and flexible enough to work across various products. It should include what’s shared across products and leave out what’s not.

> A great first step is to share your idea with our design community to see if there’s a need for your component. You can leave a comment in the [Design Systems Figma Library](https://www.figma.com/design/ficiK5ADsOuiCromUZam1E/WordPress-Design-System-(Community)?node-id=13482-72792&t=ofEJnkdwKdS5mhGX-1), or drop a question in #design-systems on Slack, where people will be happy to help.

Reviewer: In general I don't think it makes sense to qualify it as "**design** community" as these decisions usually require both dev and design input. (I don't mind if people prefer to comment on the Figma library first though, if that's more accessible to them.)

```suggestion
A great first step is to share your idea with our community to see if there’s a need for your component. You can leave a comment in the [Design Systems Figma Library](https://www.figma.com/design/ficiK5ADsOuiCromUZam1E/WordPress-Design-System-(Community)?node-id=13482-72792&t=ofEJnkdwKdS5mhGX-1), or drop a question in #design-systems on Slack, where people will be happy to help.
```

(Applies to all mentions in this file.)

Some more questions you can ask yourself if you’re thinking of creating a new component:

- Could it be used by other products or plugins?  
- Does it overlap functionally or visually with any existing components? Can we use existing components to implement an alternative solution?  
- How much effort will be required to create and maintain it?  

> - Is there a clear purpose for it?  

Reviewer 2: These questions are great. I think they might have even more impact if instead of a deterrent, they could be part of a quality assurance checklist that or something of sorts?

> - Can we introduce this change without causing breaking changes? Will we be able to keep this change and iterate on it in the future without causing future breaking changes?

Reviewer 2: Is this a breaking change? And maybe link to what makes a breaking change?

Author: I can't find WordPress guidance on what makes a breaking change — do you think I could link out to general guidance?

Alternatively, we could rewrite to something like:

“Can we make this change without causing a breaking change (something that forces users to update their setup or workflows to keep things working) and will it allow future updates without similar issues?”

Reviewer: [This](https://github.com/WordPress/gutenberg/blob/c5e8fd268328437220798137805c2fb623b25ed1/packages/README.md?plain=1#L202) is a good description:

> "Breaking Changes" - A backwards-incompatible change which requires specific attention of the impacted developers to reconcile

There's also a relevant section [here](https://github.com/WordPress/gutenberg/blob/c5e8fd268328437220798137805c2fb623b25ed1/packages/components/CONTRIBUTING.md?plain=1#L88-L104).

This flowchart can help determine if a new component is necessary:

ADD IMAGE

#### Step 2: Open a GitHub issue {#step-2}

If you’ve decided you want to create a new component, you’ll need to open a GitHub issue on Project Gutenberg. This is where we’ll be able to discuss the proposed changes and track progress.

[Click here to open a new issue](https://github.com/WordPress/gutenberg/issues/new?labels=Design+System&body=Then%2C+write+a+detailed+description%3A%0D%0A%0D%0A-+Explain+your+rationale+for+adding+this+component%0D%0A-+Detail+the+intended+behavior%0D%0A-+Clarify+if+it%E2%80%99s+a+new+component+of+a+variation+of+an+existing+one%0D%0A-+Include+a+basic+mockup+%28optional%29%0D%0A-+Include+inspiration+from+other+products+%28optional%29).

Then, write a detailed description:

- Explain your rationale for adding this component  
- Detail the intended behavior  

> - Clarify if it’s a new component of a variation of an existing one  

Reviewer 2: I think there's a typo here

- Include a basic mockup (optional)  
- Include inspiration from other products (optional)  
- Tag the [@WordPress/gutenberg-components](https://github.com/orgs/WordPress/teams/gutenberg-components) and [@WordPress/gutenberg-design](https://github.com/orgs/WordPress/teams/gutenberg-design) GitHub groups

> Add the “Design System” label to correctly categorize your work. Also, consider adding other labels depending on your needs, like Needs Design, Needs Copy Review, Needs Dev, or Needs Accessibility Feedback.

Reviewer: We could leave this out, since I don't think new contributors have privileges to add arbitrary labels anyway. More experienced contributors with privileges will label them accordingly, so new contributors mostly shouldn't have to worry about it.

Author: I see what you mean, but wouldn't it be helpful for a new contributor to get used to labeling their issues correctly?

Reviewer: They literally [cannot](https://stackoverflow.com/a/13829505) because they do not have the rights. I believe it works the same on almost all open source projects on GitHub. They gain rights when they join the [Triage team](https://developer.wordpress.org/block-editor/contributors/triage/) or meet the criteria to join the [Gutenberg team](https://developer.wordpress.org/block-editor/contributors/repository-management/#teams).

Remember, it’s unlikely that everything will be done by one person. Contribute where you can, and others will help.

**Write a detailed rationale**  
Explain how your component will add value to WPDS and the greater product ecosystem. Make sure to describe how it works and how someone can interact with it.

**Draft documentation**  
New components need development, design, and accessibility guidelines. If your change adds additional behavior or expands a component’s features, those changes will need to be documented as well.

Read through existing component documentation for inspiration. Start with a rough draft, and the community will help polish documentation.

**Provide working code**  
The component or enhancement must be built in React. For more information, see our [developer contribution guidelines](https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/README.md) and, specifically for the @wordpress/components package, [the package’s contributing guidelines](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md).

**Create a design spec**  

> You’ll need to share sizing and styling annotations for all aspects of the component, so a developer has everything they need to create the design in code. Figma’s [Dev Mode](https://help.figma.com/hc/en-us/articles/15023124644247-Guide-to-Dev-Mode) can do this automatically for you.

Reviewer 2: Are sizing and styling the only requirements for a spec?

Author: I've based myself on [existing guidance](https://wordpress.github.io/gutenberg/?path=/docs/components-contributing-guidelines--docs#next-steps) for this — I don't know if they reflect current practices :)

#### Step 4: Refinement {#step-4}

> We’re almost ready to publish your component, but you’ll still need to spend some time finetuning it. You can do the following:

Reviewer 2: ```suggestion
You’re almost ready to publish your component, but you’ll still need to spend some time finetuning it. You can do the following:
```

- First, clearly define and agree on which features will be included in the component.  
- Then, reduce the scope by removing features that lack consensus.  

> - When that’s done, you’ll have to do quality assurance to ensure that your contribution adheres to our system standards.

Reviewer 2: ```suggestion
- When that’s done, you’ll have to do quality assurance to ensure that your contribution adheres to the system standards.
```

**Quality assurance**  
Every component should be tested for quality during its development, rather than after it’s published.

Make sure to test the following aspects of your component:

- Accessibility  
- Correct semantics  

> - Key user interactions

Reviewer 2: I think it'd be cool to add a one line example to each of these aspects. For example:

Correct semantics — Does the API follow naming conventions? Are you introducing new terms?

And maybe a link to any existing documentation that helps with this.

Reviewer 2: Oh nice, there's a section on this just down below ✨

**Accessibility**  
Has your design and implementation accounted for the standards set by the [WordPress accessibility guidelines](https://make.wordpress.org/accessibility/handbook/which-questions-should-you-ask/)?

#### Accessibility Standards

Use the [WordPress Accessibility Guidelines](https://make.wordpress.org/accessibility/handbook/) as your benchmark and ensure components meet ARIA standards, work with screen readers, and are keyboard-navigable.

Important: Add the “Needs Accessibility Feedback” label to your GitHub issue early to get a review from the Accessibility team, as it can help you design inclusively from the start.

**Visual quality**  

> Does it apply visual styles correctly? Color, typography, icons, space, borders need to use appropriate variables and follow the [WordPress Style Guide](https://make.wordpress.org/docs/style-guide/).

Reviewer 2: I'd add a note that if they can't find the guidance that they're looking for to add questions or clarification in the issue itself.

Important: Add the “Needs Design Feedback” label to your GitHub issue to get a review from the Design team.

**Documentation**  
Does it have proper documentation for development, design, and accessibility?

**States and variations**  
Does it cover all necessary variations (primary, secondary, dense, etc) and states (default, hover, active, disabled, loading, etc.), within your intended scope?

**Functionality**  
Does it function as expected?

**Mobile**  

> Is it designed mobile-first, does it responsively adapt to different screen sizes, and do all touch interactions work as expected?

Reviewer 2: Is it true that we optimize for mobile use?

Author: Also based on [existing guidance](https://wordpress.github.io/gutenberg/?path=/docs/components-contributing-guidelines--docs#quality-assurance) and I don't know if it reflects current standards.

**Backward Compatibility and Deprecation**  
If your component needs updates in the future, plan changes carefully. Introduce new features in a backward-compatible way, and if you must remove or alter existing APIs, follow our guidelines for soft-deprecation and communicate timelines clearly.

By completing these steps, you ensure your component remains robust, useful, and aligned with the WordPress Design System’s mission to deliver accessible, consistent, and forward-looking UI solutions.

## Extending Existing Components

### Compatibility

The `@wordpress/components` package includes components that are relied upon by many developers across different projects. It is, therefore, very important to avoid introducing breaking changes.

In these situations, one possible approach is to "soft-deprecate" a given legacy API. This is achieved by:

1. Removing traces of the API from the docs, while still supporting it in code.  
2. Updating all places in Gutenberg that use that API.  
3. Adding deprecation warnings (only after the previous point is completed).

> When adding new components or new props to existing components, consider creating a private or experimental version until the changes are stable enough to be exposed publicly.

Reviewer 2: ```suggestion
When adding new components or new props to existing components, consider creating a private or experimental version until the changes are stable enough expose publicly.
```

Author: Adding as "to expose"

    deprecated(
      'Old styles for MyComponent',
      {
        since: '1.0',
        version: '2.0',
        hint: 'Set `__nextHasNewStyle` to true to opt into the new behavior.',
      }
    );
  }

  return <div className={__nextHasNewStyle ? 'new-style' : 'old-style'} />;
}
```

* Use clear warnings to inform developers of upcoming changes.
* Define a clear timeline for transitioning to new styles or APIs.
* Update documentation and examples to encourage adoption of the new standard.

### Context System

Use the `ContextSystemProvider` to provide values to components and `useContextSystem` to read them. Connect components to the context system with `contextConnect` so that props can be adapted automatically depending on the rendering context.

### Folder Structure

A recommended folder structure for a component:

component-name/ ├── stories │ └── index.js ├── test │ └── index.js ├── component.tsx ├── context.ts ├── hook.ts ├── index.ts ├── README.md ├── styles.ts └── types.ts\

For component families, each subcomponent gets its own folder, with shared code at the root.

### Component Versioning

If you need to introduce a radically new version of a component:

> - Attempt to keep the same API surface and just update the internal implementation.

Reviewer 2: ```suggestion
- Attempt to keep the same API surface and update the internal implementation.
```
